packages/jsii-rosetta/lib/snippet-dependencies.ts

```diff
@@ -0,0 +1,245 @@
+import * as cp from 'child_process';
+import * as fastGlob from 'fast-glob';
+import * as fs from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import { intersect } from 'semver-intersect';
+
+import { findDependencyDirectory, findUp } from './find-utils';
+import * as logging from './logging';
+import { TypeScriptSnippet, CompilationDependency } from './snippet';
+import { mkDict, formatList } from './util';
+
+/**
+ * Collect the dependencies of a bunch of snippets together in one declaration
+ *
+ * We assume here the dependencies will not conflict.
+ */
+export function collectDependencies(snippets: TypeScriptSnippet[]) {
+  const ret: Record<string, CompilationDependency> = {};
```

Comment:
FYI (don't go refactoring this!), I kind of pivoted to using `Map<String, T>` in such cases (i.e: where the key space isn't known upfront).

Reply:
I'm always going back and forth on this.

What's your motivation for picking `Map` over an object?

Reply:
There are a couple of reasons:
- `Map` typically won't lead you to prototype corruption routes, because you'll probably never be tempted to `Object.assign` on them.
  - Alternatively, you'd initialize your object with `Object.create(null)` to remove risk for future prototype pollution.
- `Map` won't create a new hidden class each time you assign a new key/property on them. This reduces memory churn, and perhaps also footprint (although hidden classes now get GC'd -- in earlier versions of V8, I reckon they'd land in "permanent gen" and stick around forever).
- `Map` can be used in `for [key, value] of map`, while for Objects you have to `Object.entries(obj)` there (which is also not `null`-safe)
- `Map` does not have some of the "surprising" behavior of objects with their keys, e.g:
  ```js
  const obj = Object.create(null);
  const map = new Map();

  obj[123] = 'FOO';
  assert(obj['123'] === 'FOO');

  map.set(123, 'FOO');
  assert(map.get('123') !== 'FOO');
  ```
- `Object` properties need to be valid property names. `Map` keys can be valid JS value.